pageserver/src/tenant/timeline.rs

```diff
@@ -1711,6 +1719,28 @@ impl Timeline {
         if let Some(rtc) = self.remote_client.as_ref() {
             rtc.schedule_layer_file_deletion(&needs_cleanup)?;
             rtc.schedule_index_upload_for_file_changes()?;
+            // This barrier orders above DELETEs before any later operations.
+            // This is critical because code executing after the barrier might
+            // create again the objects that we just scheduled for deletion.
+            // For example, if we just scheduled deletion of an image layer "from the future",
+            // later compaction might run again and re-create the same image layer.
+            // "from the future" here means an image layer whose LSN is > IndexPart::disk_consistent_lsn.
+            // "same" here means same key range and LSN.
+            // DELETEs and PUTs can race in the upload queue, so if the PUT is executed
+            // before the DELETE, the DELETE deletes the wrong layer.
```

Comment:
```suggestion
            // before the DELETE, the DELETE is final, against our expectations of eventual re-creation.
            // This leads to index-part.json mentioning layers that are not present in S3.
```

If the two layers are the same (and they are, they contain the same data after all), then "deletes the wrong layer" sounds wrong to me.